File: src/ipc/permissions.ts
import type { Role } from "../stores/community.store";

// Permission bit constants — mirrors Rust's `rekindle_protocol::dht::community::permissions_v2`.
// Uses Discord-aligned bit positions. BigInt is used throughout for correct 64-bit bitwise ops.

// ── General ──
export const CREATE_INSTANT_INVITE = 1n << 0n;
export const KICK_MEMBERS = 1n << 1n;
export const BAN_MEMBERS = 1n << 2n;
export const ADMINISTRATOR = 1n << 3n;
export const MANAGE_CHANNELS = 1n << 4n;
export const MANAGE_COMMUNITY = 1n << 5n;

// ── Text ──
export const ADD_REACTIONS = 1n << 6n;
export const VIEW_AUDIT_LOG = 1n << 7n;
export const PRIORITY_SPEAKER = 1n << 8n;
export const STREAM = 1n << 9n;
export const VIEW_CHANNEL = 1n << 10n;
export const SEND_MESSAGES = 1n << 11n;
// bit 12 reserved
export const MANAGE_MESSAGES = 1n << 13n;
export const EMBED_LINKS = 1n << 14n;
export const ATTACH_FILES = 1n << 15n;
export const READ_MESSAGE_HISTORY = 1n << 16n;
export const MENTION_EVERYONE = 1n << 17n;
export const USE_EXTERNAL_EMOJIS = 1n << 18n;
// bit 19 reserved

// ── Voice ──
export const CONNECT = 1n << 20n;
export const SPEAK = 1n << 21n;
export const MUTE_MEMBERS = 1n << 22n;
export const DEAFEN_MEMBERS = 1n << 23n;
export const MOVE_MEMBERS = 1n << 24n;
export const USE_VAD = 1n << 25n;

// ── Membership ──
export const CHANGE_NICKNAME = 1n << 26n;
export const MANAGE_NICKNAMES = 1n << 27n;
export const MANAGE_ROLES = 1n << 28n;
// bits 29-32 reserved

// ── Events ──
export const MANAGE_EVENTS = 1n << 33n;

// ── Threads ──
export const MANAGE_THREADS = 1n << 34n;
export const CREATE_PUBLIC_THREADS = 1n << 35n;
export const CREATE_PRIVATE_THREADS = 1n << 36n;
// bits 37-39 reserved

// ── Moderation ──
export const MODERATE_MEMBERS = 1n << 40n;
// bits 41-42 reserved

// ── Advanced ──
export const USE_APPLICATION_COMMANDS = 1n << 43n;
export const REQUEST_TO_SPEAK = 1n << 44n;
export const MANAGE_GUILD_EXPRESSIONS = 1n << 45n;
export const MANAGE_WEBHOOKS = 1n << 46n;
// bits 47-48 reserved
export const CREATE_GUILD_EXPRESSIONS = 1n << 49n;
// bits 50-52 reserved
export const SEND_VOICE_MESSAGES = 1n << 53n;
// bits 54-55 reserved
export const SEND_POLLS = 1n << 56n;
export const VIEW_CREATOR_MONETIZATION_ANALYTICS = 1n << 57n;

// All defined permission bits OR'd together — mirrors Rust's `Permissions::all()`.
export const ALL_PERMISSIONS =
  CREATE_INSTANT_INVITE | KICK_MEMBERS | BAN_MEMBERS | ADMINISTRATOR
  | MANAGE_CHANNELS | MANAGE_COMMUNITY | ADD_REACTIONS | VIEW_AUDIT_LOG
  | PRIORITY_SPEAKER | STREAM | VIEW_CHANNEL | SEND_MESSAGES
  | MANAGE_MESSAGES | EMBED_LINKS | ATTACH_FILES | READ_MESSAGE_HISTORY
  | MENTION_EVERYONE | USE_EXTERNAL_EMOJIS | CONNECT | SPEAK
  | MUTE_MEMBERS | DEAFEN_MEMBERS | MOVE_MEMBERS | USE_VAD
  | CHANGE_NICKNAME | MANAGE_NICKNAMES | MANAGE_ROLES
  | MANAGE_EVENTS | MANAGE_THREADS | CREATE_PUBLIC_THREADS | CREATE_PRIVATE_THREADS
  | MODERATE_MEMBERS | USE_APPLICATION_COMMANDS | REQUEST_TO_SPEAK
  | MANAGE_GUILD_EXPRESSIONS | MANAGE_WEBHOOKS | CREATE_GUILD_EXPRESSIONS
  | SEND_VOICE_MESSAGES | SEND_POLLS | VIEW_CREATOR_MONETIZATION_ANALYTICS;

/**
 * Check if a permission set includes a specific permission.
 * Administrators bypass all checks.
 */
export function hasPermission(perms: bigint, required: bigint): boolean {
  if ((perms & ADMINISTRATOR) !== 0n) return true;
  return (perms & required) === required;
}

/**
 * Check if permissions include the ADMINISTRATOR flag.
 */
export function isAdministrator(perms: bigint): boolean {
  return (perms & ADMINISTRATOR) !== 0n;
}

/**
 * Calculate base permissions for a member from their role IDs.
 * ORs together all role permissions for the member's roles.
 * This is a simplified client-side version for UI gating.
 *
 * Role.permissions is a string (serialized from Rust u64 to avoid JS Number precision loss).
 * Internally converts to BigInt for correct 64-bit bitwise operations.
 */
export function calculateBasePermissions(
  roleIds: number[],
  allRoles: Role[],
): bigint {
  let perms = 0n;
  for (const id of roleIds) {
    const role = allRoles.find((r) => r.id === id);
    if (role) {
      perms |= BigInt(role.permissions);
    }
  }
  return perms;
}

/**
 * Check if the user with the given highest role position can manage
 * a target with the given role position (must be strictly higher).
 */
export function canManageRole(
  myHighestPosition: number,
  targetPosition: number,
): boolean {
  return myHighestPosition > targetPosition;
}

/**
 * Get the highest role position from a set of role IDs.
 */
export function highestPosition(
  roleIds: number[],
  allRoles: Role[],
): number {
  let max = -1;
  for (const id of roleIds) {
    const role = allRoles.find((r) => r.id === id);
    if (role && role.position > max) {
      max = role.position;
    }
  }
  return max;
}

/**
 * Toggle a permission bit in a permission set.
 */
export function togglePermBit(current: bigint, bit: bigint): bigint {
  return current ^ bit;
}

/**
 * All permission constants grouped by category, for use in the role editor UI.
 */
export const PERMISSION_CATEGORIES = [
  {
    name: "General",
    permissions: [
      { key: "CREATE_INSTANT_INVITE", label: "Create Invite", value: CREATE_INSTANT_INVITE },
      { key: "MANAGE_CHANNELS", label: "Manage Channels", value: MANAGE_CHANNELS },
      { key: "MANAGE_COMMUNITY", label: "Manage Community", value: MANAGE_COMMUNITY },
      { key: "VIEW_AUDIT_LOG", label: "View Audit Log", value: VIEW_AUDIT_LOG },
      { key: "MANAGE_ROLES", label: "Manage Roles", value: MANAGE_ROLES },
      { key: "MANAGE_NICKNAMES", label: "Manage Nicknames", value: MANAGE_NICKNAMES },
      { key: "CHANGE_NICKNAME", label: "Change Nickname", value: CHANGE_NICKNAME },
      { key: "ADMINISTRATOR", label: "Administrator", value: ADMINISTRATOR },
    ],
  },
  {
    name: "Text",
    permissions: [
      { key: "VIEW_CHANNEL", label: "View Channels", value: VIEW_CHANNEL },
      { key: "SEND_MESSAGES", label: "Send Messages", value: SEND_MESSAGES },
      { key: "MANAGE_MESSAGES", label: "Manage Messages", value: MANAGE_MESSAGES },
      { key: "EMBED_LINKS", label: "Embed Links", value: EMBED_LINKS },
      { key: "ATTACH_FILES", label: "Attach Files", value: ATTACH_FILES },
      { key: "READ_MESSAGE_HISTORY", label: "Read History", value: READ_MESSAGE_HISTORY },
      { key: "ADD_REACTIONS", label: "Add Reactions", value: ADD_REACTIONS },
      { key: "MENTION_EVERYONE", label: "Mention Everyone", value: MENTION_EVERYONE },
      { key: "USE_EXTERNAL_EMOJIS", label: "External Emojis", value: USE_EXTERNAL_EMOJIS },
    ],
  },
  {
    name: "Voice",
    permissions: [
      { key: "CONNECT", label: "Connect", value: CONNECT },
      { key: "SPEAK", label: "Speak", value: SPEAK },
      { key: "STREAM", label: "Stream", value: STREAM },
      { key: "MUTE_MEMBERS", label: "Mute Members", value: MUTE_MEMBERS },
      { key: "DEAFEN_MEMBERS", label: "Deafen Members", value: DEAFEN_MEMBERS },
      { key: "MOVE_MEMBERS", label: "Move Members", value: MOVE_MEMBERS },
      { key: "USE_VAD", label: "Use Voice Detection", value: USE_VAD },
      { key: "PRIORITY_SPEAKER", label: "Priority Speaker", value: PRIORITY_SPEAKER },
      { key: "REQUEST_TO_SPEAK", label: "Request to Speak", value: REQUEST_TO_SPEAK },
    ],
  },
  {
    name: "Threads",
    permissions: [
      { key: "MANAGE_THREADS", label: "Manage Threads", value: MANAGE_THREADS },
      { key: "CREATE_PUBLIC_THREADS", label: "Create Public Threads", value: CREATE_PUBLIC_THREADS },
      { key: "CREATE_PRIVATE_THREADS", label: "Create Private Threads", value: CREATE_PRIVATE_THREADS },
    ],
  },
  {
    name: "Events",
    permissions: [
      { key: "MANAGE_EVENTS", label: "Manage Events", value: MANAGE_EVENTS },
    ],
  },
  {
    name: "Moderation",
    permissions: [
      { key: "KICK_MEMBERS", label: "Kick Members", value: KICK_MEMBERS },
      { key: "BAN_MEMBERS", label: "Ban Members", value: BAN_MEMBERS },
      { key: "MODERATE_MEMBERS", label: "Moderate Members", value: MODERATE_MEMBERS },
    ],
  },
  {
    name: "Advanced",
    permissions: [
      { key: "USE_APPLICATION_COMMANDS", label: "Application Commands", value: USE_APPLICATION_COMMANDS },
      { key: "MANAGE_GUILD_EXPRESSIONS", label: "Manage Expressions", value: MANAGE_GUILD_EXPRESSIONS },
      { key: "CREATE_GUILD_EXPRESSIONS", label: "Create Expressions", value: CREATE_GUILD_EXPRESSIONS },
      { key: "MANAGE_WEBHOOKS", label: "Manage Webhooks", value: MANAGE_WEBHOOKS },
      { key: "SEND_VOICE_MESSAGES", label: "Voice Messages", value: SEND_VOICE_MESSAGES },
      { key: "SEND_POLLS", label: "Send Polls", value: SEND_POLLS },
      { key: "VIEW_CREATOR_MONETIZATION_ANALYTICS", label: "View Analytics", value: VIEW_CREATOR_MONETIZATION_ANALYTICS },
    ],
  },
] as const;

File: crates/rekindle-transport/src/community/permissions.rs
//! Permission bitmask constants, preset roles, and the full 8-step
//! Discord-aligned permission computation algorithm.

use serde::{Deserialize, Serialize};

use crate::payload::dht_types::RoleEntry;

/// The @everyone role always has ID 0.
pub const ROLE_EVERYONE_ID: u32 = 0;

/// Permission overwrite for a channel, targeting either a role or a member.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PermissionOverwrite {
    pub target_type: OverwriteType,
    pub target_id: String,
    pub allow: u64,
    pub deny: u64,
}

/// Whether a permission overwrite targets a role or a member.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum OverwriteType {
    Role,
    Member,
}

bitflags::bitflags! {
    /// Type-safe permission bitmask. Discord-aligned bit positions.
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    pub struct Permissions: u64 {
        const CREATE_INSTANT_INVITE     = 1 << 0;
        const KICK_MEMBERS              = 1 << 1;
        const BAN_MEMBERS               = 1 << 2;
        const ADMINISTRATOR             = 1 << 3;
        const MANAGE_CHANNELS           = 1 << 4;
        const MANAGE_COMMUNITY          = 1 << 5;
        const ADD_REACTIONS             = 1 << 6;
        const VIEW_AUDIT_LOG            = 1 << 7;
        const PRIORITY_SPEAKER          = 1 << 8;
        const STREAM                    = 1 << 9;
        const VIEW_CHANNEL              = 1 << 10;
        const SEND_MESSAGES             = 1 << 11;
        const MANAGE_MESSAGES           = 1 << 13;
        const EMBED_LINKS               = 1 << 14;
        const ATTACH_FILES              = 1 << 15;
        const READ_MESSAGE_HISTORY      = 1 << 16;
        const MENTION_EVERYONE          = 1 << 17;
        const USE_EXTERNAL_EMOJIS       = 1 << 18;
        const CONNECT                   = 1 << 20;
        const SPEAK                     = 1 << 21;
        const MUTE_MEMBERS              = 1 << 22;
        const DEAFEN_MEMBERS            = 1 << 23;
        const MOVE_MEMBERS              = 1 << 24;
        const USE_VAD                   = 1 << 25;
        const CHANGE_NICKNAME           = 1 << 26;
        const MANAGE_NICKNAMES          = 1 << 27;
        const MANAGE_ROLES              = 1 << 28;
        const MANAGE_EVENTS             = 1 << 33;
        const MANAGE_THREADS            = 1 << 34;
        const CREATE_PUBLIC_THREADS     = 1 << 35;
        const CREATE_PRIVATE_THREADS    = 1 << 36;
        const MODERATE_MEMBERS          = 1 << 40;
        const USE_APPLICATION_COMMANDS  = 1 << 43;
        const REQUEST_TO_SPEAK          = 1 << 44;
        const MANAGE_GUILD_EXPRESSIONS  = 1 << 45;
        const MANAGE_WEBHOOKS           = 1 << 46;
        const CREATE_GUILD_EXPRESSIONS  = 1 << 49;
        const SEND_VOICE_MESSAGES       = 1 << 53;
        const SEND_POLLS                = 1 << 56;
        const VIEW_CREATOR_MONETIZATION_ANALYTICS = 1 << 57;
    }
}

impl serde::Serialize for Permissions {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        self.bits().serialize(serializer)
    }
}

impl<'de> serde::Deserialize<'de> for Permissions {
    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        Ok(Self::from_bits_truncate(u64::deserialize(deserializer)?))
    }
}

impl Permissions {
    pub fn has(self, required: Self) -> bool {
        self.contains(Self::ADMINISTRATOR) || self.contains(required)
    }
}

/// Check permission from raw u64 values (backward-compatible wrapper).
pub fn has_permission(member_permissions: u64, required: u64) -> bool {
    Permissions::from_bits_truncate(member_permissions)
        .has(Permissions::from_bits_truncate(required))
}

// ── Default permission presets ───────────────────────────────────────

pub fn everyone_default() -> Permissions {
    Permissions::VIEW_CHANNEL | Permissions::READ_MESSAGE_HISTORY | Permissions::CONNECT
        | Permissions::SEND_MESSAGES | Permissions::SPEAK | Permissions::ADD_REACTIONS
        | Permissions::EMBED_LINKS | Permissions::ATTACH_FILES | Permissions::USE_EXTERNAL_EMOJIS
        | Permissions::USE_VAD | Permissions::CHANGE_NICKNAME | Permissions::SEND_VOICE_MESSAGES
        | Permissions::SEND_POLLS
}

pub fn member_default() -> Permissions {
    everyone_default() | Permissions::CREATE_INSTANT_INVITE
        | Permissions::CREATE_PUBLIC_THREADS | Permissions::USE_APPLICATION_COMMANDS
}

pub fn moderator_default() -> Permissions {
    member_default() | Permissions::KICK_MEMBERS | Permissions::MANAGE_MESSAGES
        | Permissions::MUTE_MEMBERS | Permissions::DEAFEN_MEMBERS
        | Permissions::MODERATE_MEMBERS | Permissions::MANAGE_EVENTS | Permissions::MANAGE_THREADS
}

pub fn admin_default() -> Permissions {
    moderator_default() | Permissions::ADMINISTRATOR | Permissions::MANAGE_CHANNELS
        | Permissions::MANAGE_ROLES | Permissions::BAN_MEMBERS | Permissions::VIEW_AUDIT_LOG
        | Permissions::MANAGE_NICKNAMES | Permissions::MANAGE_COMMUNITY
        | Permissions::MANAGE_GUILD_EXPRESSIONS | Permissions::MANAGE_WEBHOOKS
}

pub fn owner_default() -> Permissions { Permissions::all() }

/// Full 8-step Discord permission algorithm.
pub fn calculate_permissions(
    member_role_ids: &[u32],
    all_roles: &[RoleEntry],
    channel_overwrites: &[PermissionOverwrite],
    member_pseudonym: &str,
    is_owner: bool,
    timeout_until: Option<u64>,
) -> Permissions {
    if is_owner { return Permissions::all(); }

    // Step 1: @everyone base
    let everyone_perms = all_roles.iter()
        .find(|r| r.id == ROLE_EVERYONE_ID)
        .map_or(Permissions::empty(), |r| Permissions::from_bits_truncate(r.permissions));

    // Step 2: OR all assigned roles
    let mut base = everyone_perms;
    for rid in member_role_ids {
        if *rid != ROLE_EVERYONE_ID {
            if let Some(role) = all_roles.iter().find(|r| r.id == *rid) {
                base |= Permissions::from_bits_truncate(role.permissions);
            }
        }
    }

    // Step 3: ADMINISTRATOR bypass
    if base.contains(Permissions::ADMINISTRATOR) { return Permissions::all(); }

    let mut perms = base;

    if !channel_overwrites.is_empty() {
        let eid = ROLE_EVERYONE_ID.to_string();
        // Step 4: @everyone channel overwrite
        for ow in channel_overwrites {
            if ow.target_type == OverwriteType::Role && ow.target_id == eid {
                perms &= !Permissions::from_bits_truncate(ow.deny);
                perms |= Permissions::from_bits_truncate(ow.allow);
            }
        }
        // Step 5: Role-specific overwrites
        let (mut ra, mut rd) = (Permissions::empty(), Permissions::empty());
        for ow in channel_overwrites {
            if ow.target_type == OverwriteType::Role {
                if let Ok(rid) = ow.target_id.parse::<u32>() {
                    if rid != ROLE_EVERYONE_ID && member_role_ids.contains(&rid) {
                        ra |= Permissions::from_bits_truncate(ow.allow);
                        rd |= Permissions::from_bits_truncate(ow.deny);
                    }
                }
            }
        }
        perms &= !rd;
        perms |= ra;
        // Step 6: Member-specific overwrite
        for ow in channel_overwrites {
            if ow.target_type == OverwriteType::Member && ow.target_id == member_pseudonym {
                perms &= !Permissions::from_bits_truncate(ow.deny);
                perms |= Permissions::from_bits_truncate(ow.allow);
            }
        }
    }

    // Step 7: Timeout strips write/voice
    if let Some(until) = timeout_until {
        if rekindle_utils::timestamp_secs() < until {
            perms &= !(Permissions::SEND_MESSAGES | Permissions::ADD_REACTIONS
                | Permissions::SPEAK | Permissions::STREAM
                | Permissions::CREATE_INSTANT_INVITE | Permissions::SEND_VOICE_MESSAGES
                | Permissions::SEND_POLLS);
        }
    }

    // Step 8: No VIEW_CHANNEL → deny all
    if !perms.contains(Permissions::VIEW_CHANNEL) { return Permissions::empty(); }

    perms
}
